lib/plaid/categories.ex

```diff
@@ -24,12 +26,34 @@ defmodule Plaid.Categories do
   @doc """
   Gets all categories.
   """
-  @spec get(config | nil) :: {:ok, Plaid.Categories.t()} | {:error, Plaid.Error.t()}
+  @spec get(config) :: {:ok, Plaid.Categories.t()} | error
   def get(config \\ %{}) do
-    config = Map.drop(config, [:public_key, :client_id, :secret])
-    endpoint = "#{@endpoint}/get"
+    c = config[:client] || Plaid
+
+    Request
+    |> struct(method: :post, endpoint: "categories/get")
+    |> Request.add_metadata(config)
+    |> c.send_request(Client.new(config))
+    |> c.handle_response()
+    |> case do
+      {:ok, body} ->
+        {:ok, map_categories(body)}
+
+      {:error, _} = error ->
+        error
+    end
```

Comment:
Or inline the call of `Poison.Decode.transform` instead of adding `map_categories` at that point

Reply:
I like this. What about taking this concept and adding it to `Plaid.handle_response/1`? I could replace that callback with a `Plaid.handle_response/2` that takes the mapping function as the second argument as you do in `map_ok`:

```elixir
@callback handle_response({:ok, Tesla.Env.t()} | {:error, any()}, (any() -> any())) :: {:ok, term} | {:error, Plaid.Error.t()} | {:error, any()}

# Implementation
def handle_response({:ok, %Tesla.Env{status: status} = env}, map_fun) when status in 200..299 do
  {:ok, map_fun.(env.body)}
end

def handle_response({:ok, %Tesla.Env{} = env}, _map_fun) do
  {:error, Poison.Decode.transform(env.body, %{as: %Plaid.Error{}})}
end

def handle_response({:error, _reason} = error, _map_fun) do
  error
end
```

It could then be used in the module like so:
```elixir
Result
|> struct(method: :post, endpoint: "categories/get")
|> Request.add_metadata(config)
|> c.send_request(Client.new(config))
|> c.handle_response(map_categories)
```

Reply:
I'm doing something similar here already. But I think your solution is more elegant. https://github.com/wfgilman/plaid-elixir/blob/afd567ef8aaeba09728556a366f4cd31e761bcc0/lib/plaid/item.ex#L91-L111

Reply:
Implemented in the latest commit 🙂